simulation.py

```diff
@@ -1,293 +1,524 @@
+# with a simple placeholder graph/plot, threading, and signalling  between the thread and 
+# the main window.
+# 12/1/20 Adding a data storage area to share between the SimulationThread and MainWindow thread
+# which incorporates a mutex mechanism (QReadWriteLock) to allow coordinating sharing of the
+# data which MainWindow will be consuming (reading).
+# 12/52/20 Manual merge in branch 'one-lock-rules-them-all' simulation code with the QThread
+# architecture framed in from the previous versions of this branch
+
 # USE ONLY SI UNITS
 import sys
 import time
 import logging
-import threading
-from PyQt5 import QtWidgets
+from PyQt5 import QtGui, QtCore
+from PyQt5.QtCore import *
+from PyQt5.QtGui import *
+from PyQt5.QtWidgets import *
 import pyqtgraph as pg
+import numpy as np
 import cProfile
-
-
+from datastore import (DataStore, RacingSimulationResults)
+from logging_config import configure_logging
 from physics_equations import (max_negative_power_physics_simulation,
                                max_positive_power_physics_simulation,
                                constrained_velocity_physics_simulation,
                                )
 from electric_car_properties import ElectricCarProperties
 from track_properties import TrackProperties
-
 from logging_config import configure_logging
-from datastore import (DataStore,
-                       RacingSimulationResults,
-                       )
 
 logger = logging.getLogger(__name__)
 
 
-def racing_simulation(data_store):
-    """Function accepts a car and a track and executes
-    a simulation to ouput critical metrics related
-    to battery life and track speed.
-
-    Args:
-        data_store (DataStore): Thread safe storage for all simulation data
-
-    Returns:
-        Nothing (all data saved in the datastore)
-
-    """
-    results = RacingSimulationResults()
-
-    lap_velocity_simulation(data_store)
-    # only calculate results if the simulation ran through without an interruption
-    if not data_store.exit_event.is_set():
-        lap_results = data_store.get_lap_results()
-        car = data_store.get_car_properties()
-        results.laps_per_pit_stop = car["battery_capacity"] / lap_results.motor_energy_list[-1]
-        results.lap_time = lap_results.end_velocity
-        results.lap_results = lap_results
-        data_store.set_race_results(results)
-
-
-def lap_velocity_simulation(data_store):
-    """Function calculates the velocity profile of a car with
-    car_properties on a track with track_properties. The car
-    starts with an ititial velocity of initial_velocity.
 
-    Args:
-        data_store (DataStore): Thread safe storage for all simulation data
+class MainWindow(QWidget):
 
-    Returns:
-        Nothing (all data saved in the datastore)
+    def __init__(self, *args, **kwargs):
+        QWidget.__init__(self, parent = None)
+        print('Window: __init__')
```

Comment:
Eventually I'd like all prints to become logging calls, we can define custom levels of logging and turn them on/off with arguments, for now #TODO this